sw-emulator/lib/periph/src/hash_sha3.rs

```diff
@@ -0,0 +1,793 @@
+            0x16 => CmdType::Done,
+            _ => panic!("Invalid command type"),
+        }
+    }
+}
+
+enum Endianness {
+    Little = 0x0,
+    Big = 0x1,
+}
+
+impl From<Endianness> for u32 {
+    fn from(mode: Endianness) -> Self {
+        mode as Self
+    }
+}
+
+impl From<u32> for Endianness {
+    fn from(value: u32) -> Self {
+        match value {
+            0x0 => Endianness::Little,
+            0x1 => Endianness::Big,
+            _ => panic!("Invalid endianness"),
+        }
+    }
+}
+
+fn u32_to_u8_le(input: &[u32]) -> Vec<u8> {
+    input.iter().flat_map(|n| n.to_le_bytes()).collect()
+}
+
+fn u32_to_u8_be(input: &[u32]) -> Vec<u8> {
+    input.iter().flat_map(|n| n.to_be_bytes()).collect()
+}
+
+fn digest_to_state(input: [u8; 200]) -> [u32; 50] {
+    let mut output = [0u32; 50];
+    for (i, chunk) in input.chunks(4).enumerate() {
+        let mut array = [0u8; 4];
+        array.copy_from_slice(chunk);
+        output[i] = u32::from_le_bytes(array);
+    }
+
+    output
+}
+
+register_bitfields! [
+    u32,
+
+    /// Alert Test Register Fields
+    AlertTest [
+        RECOV_OPERATION_ERR OFFSET(0) NUMBITS(1) [],
+        FATAL_FAULT_ERR OFFSET(1) NUMBITS(1) [],
+        RSVD OFFSET(2) NUMBITS(30) [],
+    ],
+
+    /// Cfg Write Enable Register Fields
+    CfgRegWen [
+        EN OFFSET(0) NUMBITS(1) [],
+        RSVD OFFSET(1) NUMBITS(31) [],
+    ],
+
+    /// Cfg Shadowed Register Fields
+    CfgShadowed [
+        RSVD1 OFFSET(0) NUMBITS(1) [],
+        KSTRENGTH OFFSET(1) NUMBITS(3) [],
+        MODE OFFSET(4) NUMBITS(2) [],
+        RSVD2 OFFSET(6) NUMBITS(2) [],
+        MSG_ENDIANNESS OFFSET(8) NUMBITS(1) [],
+        STATE_ENDIANNESS OFFSET(9) NUMBITS(1) [],
+        RSVD3 OFFSET(10) NUMBITS(22) [],
+    ],
+
+    /// Command Register Fields
+    Cmd [
+        CMD OFFSET(0) NUMBITS(6) [],
+        RSVD1 OFFSET(6) NUMBITS(4) [],
+        ERR_PROCESSED OFFSET(10) NUMBITS(1) [],
+        RSVD2 OFFSET(11) NUMBITS(21) [],
+    ],
+
+    /// Status Register Fields
+    Status [
+        SHA3_IDLE OFFSET(0) NUMBITS(1) [],
+        SHA3_ABSORB OFFSET(1) NUMBITS(1) [],
+        SHA3_SQUEEZE OFFSET(2) NUMBITS(1) [],
+        RSVD1 OFFSET(3) NUMBITS(5) [],
+        FIFO_DEPTH OFFSET(8) NUMBITS(5) [],
+        RSVD2 OFFSET(13) NUMBITS(1) [],
+        FIFO_EMPTY OFFSET(14) NUMBITS(1) [],
+        FIFO_FULL OFFSET(15) NUMBITS(1) [],
+        ALERT_FATAL_FAULT OFFSET(16) NUMBITS(1) [],
+        ALERT_RECOV_CTRL_UPDATE_ERR OFFSET(17) NUMBITS(1) [],
+        RSVD3 OFFSET(18) NUMBITS(14) [],
+    ],
+
+    /// Err Code Register Fields
+    ErrCode [
+        ERR_CODE OFFSET(0) NUMBITS(32) [],
+    ],
+];
+
+const SHA3_STATE_MEMORY_SIZE: usize = 1600 / 32;
+const SHA3_MSG_FIFO_SIZE: usize = 2048 / 32;
+// TODO: I think this value is wrong.
+const SHA3_MSG_FIFO_MAX_DEPTH: usize = 32;
+
+/// SHA3 Peripheral
+#[derive(Bus)]
+#[poll_fn(poll)]
+#[warm_reset_fn(warm_reset)]
+#[update_reset_fn(update_reset)]
+pub struct HashSha3 {
```

Comment:
This also needs to be added to the `root_bus.rs` so it will be available to the RISC-V core.